### PR TITLE
Rework the CC with new features

### LINKED ISSUE #90: Individual CC elections & terms
Instead of electing the entire CC at once, we want to be able to hold elections that modify the CC in some ways. Specifically, we should be able to

- remove a set of cold keys,
- add a set of cold keys (each with a term length) and
- set a new threshold

in one election. Additionally, a cold key should be able to resign by publishing a `ccRegHot` certificate that contains `nothing` as a hot key. Resignation should have the same semantics as expiration of the term, which is that those should still count towards the size of the committee for voting, but not towards the size for `minCCSize`. I.e. if there are less active keys than `minCCSize` we go straight to 'no confidence'.

### PR DESCRIPTION
**Pull request by WhatisRT**
Closes #90

### CODE REVIEW

File: src/Ledger/GovernanceActions.lagda [hunk 1]
```diff
@@ -47,22 +48,31 @@ record Anchor : Set where
...
 data GovAction : Set where
-  NoConfidence     :                        GovAction
-  NewCommittee     : ℙ KeyHash → ℚ        → GovAction
-  NewConstitution  : DocHash              → GovAction
-  TriggerHF        : ProtVer              → GovAction
-  ChangePParams    : UpdateT → PPHash     → GovAction
-  TreasuryWdrl     : (Credential ↛ Coin)  → GovAction
-  Info             :                        GovAction
+  NoConfidence     :                                    GovAction
+  NewCommittee     : KeyHash ↛ Epoch → ℙ KeyHash → ℚ → GovAction
```
Comment (Soupstraw) [inline src/Ledger/GovernanceActions.lagda:52]: Why does it have both a map and a set for the hashes?

File: src/Ledger/GovernanceActions.lagda [hunk 2]
```diff
@@ -115,10 +126,12 @@ data _⊢_⇀⦇_,ENACT⦈_ where
...
 
   -- TODO: add hashes to everything except withdrawals & no ops
-  Enact-NoConf    : _ ⊢ s ⇀⦇ NoConfidence        ,ENACT⦈  record s { cc = nothing }
-  Enact-NewComm   : _ ⊢ s ⇀⦇ NewCommittee mem q  ,ENACT⦈  record s { cc = just (mem , q) }
-  Enact-NewConst  : _ ⊢ s ⇀⦇ NewConstitution dh  ,ENACT⦈  record s { constitution = dh }
-  Enact-HF        : _ ⊢ s ⇀⦇ TriggerHF v         ,ENACT⦈  record s { pv = v }
+  Enact-NoConf    : _ ⊢ s ⇀⦇ NoConfidence            ,ENACT⦈  record s { cc = nothing }
+  Enact-NewComm   : _ ⊢ s ⇀⦇ NewCommittee new rem q  ,ENACT⦈ let
+    old = maybe proj₁ ∅ᵐ (EnactState.cc s)
+    in record s { cc = just ((new ∪ᵐˡ old) ∣ rem , q) }
```
Comment (Soupstraw) [inline src/Ledger/GovernanceActions.lagda:132]:
```suggestion
    in record s { cc = just ((new ∪ᵐˡ old) ∣ rem ᶜ , q) }
```

File: src/Ledger/Ratify.lagda [hunk 3]
```diff
@@ -35,6 +37,7 @@ record StakeDistrs : Set where
 record RatifyEnv : Set where
   field stakeDistrs   : StakeDistrs
         currentEpoch  : Epoch
+        ccHotKeys     : KeyHash ↛ Maybe KeyHash
```
Comment (Soupstraw) [inline src/Ledger/Ratify.lagda:40]: Why does this map to a `Maybe KeyHash`?

File: src/Ledger/Ratify.lagda [hunk 4]
```diff
@@ -108,15 +111,38 @@ restrictedDists coins rank dists = dists -- record dists { drepStakeDistr = rest
...
-votedHashes : Vote → ((GovRole × Credential) ↛ Vote) → GovRole → ℙ Credential
+CCData : Set
+CCData = Maybe (KeyHash ↛ Epoch × R.ℚ)
+
+module _ (ce : Epoch) (ccHotKeys : KeyHash ↛ Maybe KeyHash)
+         (cc : CCData) (votes : (GovRole × Credential) ↛ Vote) where
+
+  actualCCVote : KeyHash → Epoch → Vote
+  actualCCVote kh e = case ⌊ ce ≤ᵉ? e ⌋ ,′ lookupᵐ? ccHotKeys kh ⦃ _ ∈? _ ⦄ of λ where
+    (false , just (just hk)) → maybe′ id Vote.no $ lookupᵐ? votes (CC , (inj₁ hk)) ⦃ _ ∈? _ ⦄
```
Comment (Soupstraw) [inline src/Ledger/Ratify.lagda:122]:
```suggestion
    (true , just (just hk)) → maybe′ id Vote.no $ lookupᵐ? votes (CC , (inj₁ hk)) ⦃ _ ∈? _ ⦄
```
If `ce` is the current epoch and `e` is the end of term, then `ce` should be less than `e` for the vote to be valid.

File: src/Ledger/Tally.lagda [hunk 5]
```diff
@@ -64,19 +63,21 @@ addAction s e aid c addr a = insert s aid record
   { votes = ∅ᵐ ; deposit = c ; returnAddr = addr ; expiresIn = e ; action = a }
 
 -- x is the anchor in those two cases, which we don't do anything with
-data _⊢_⇀⦇_,TALLY'⦈_ : TallyEnv → TallyState → GovProcedure → TallyState → Set where
-  TallyVote : ∀ {x} → let open TallyEnv Γ in
+data _⊢_⇀⦇_,TALLY'⦈_ : TallyEnv × ℕ → TallyState → GovVote ⊎ GovProposal → TallyState → Set where
```
Comment (Soupstraw) [inline src/Ledger/Tally.lagda:66]: Why not just put the governance action index into `TallyEnv`?

Comment (Soupstraw) [reply src/Ledger/Tally.lagda:66]: I got my answer in another PR

File: src/Ledger/Ratify.lagda (hunk 3 above)
Comment (WhatisRT) [reply src/Ledger/Ratify.lagda:40]: This is the mechanism for resigning. If you put a `nothing` in there, you have resigned & you can't un-resign anymore.

**Review by Soupstraw — APPROVED** [review]
LGTM

**Pull request merged**